### ISSUE TITLE
R8 breaks Workers

### ISSUE THREAD

**Issue by SalvatoreT** [square/workflow#626]
If you set `minifyEnabled` to `true` in your `build.gradle`, the second time you try starting a worker, you get an exception.

To repro, enabled `minifyEnabled` in the dungeon app (`samples/dungeon/app/build.gradle`).

```
buildTypes {
  debug {
    minifyEnabled true
    proguardFiles getDefaultProguardFile('proguard-android-optimize.txt'), 'proguard-rules.pro'
  }
}
```

and make a proguard file (`samples/dungeon/app/proguard-rules.pro`) that doesn't strip out the symbols

```
-dontobfuscate
```

Then when you run into a ghost, you get this crash.

```
E/AndroidRuntime: FATAL EXCEPTION: main
    Process: com.squareup.sample.dungeon, PID: 14615
    java.lang.IllegalStateException: Already resumed
        at kotlinx.coroutines.selects.SelectBuilderImpl.resumeWith(Select.kt:458)
        at kotlinx.coroutines.selects.SelectBuilderImpl.handleBuilderException(Select.kt:309)
        at com.squareup.workflow.internal.RealWorkflowLoop$runWorkflowLoop$2.invokeSuspend(WorkflowLoop.kt:135)
        at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
        at kotlinx.coroutines.selects.SelectBuilderImpl.resumeWith(Select.kt:254)
        at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:46)
        at kotlinx.coroutines.DispatchedTask.run(Dispatched.kt:241)
        at kotlinx.coroutines.EventLoop.processUnconfinedEvent(EventLoop.common.kt:68)
        at kotlinx.coroutines.DispatchedContinuation.resumeWith(Dispatched.kt:354)
        at kotlin.coroutines.ContinuationKt.startCoroutine(Continuation.kt:128)
        at kotlinx.coroutines.channels.AbstractChannel$ReceiveSelect.completeResumeReceive(AbstractChannel.kt:936)
        at kotlinx.coroutines.channels.AbstractSendChannel.offerInternal(AbstractChannel.kt:51)
        at kotlinx.coroutines.channels.AbstractSendChannel.offer(AbstractChannel.kt:160)
        at kotlinx.coroutines.channels.AbstractSendChannel.send(AbstractChannel.kt:146)
        at kotlinx.coroutines.channels.ChannelCoroutine.send$suspendImpl(Unknown Source:2)
        at kotlinx.coroutines.channels.ChannelCoroutine.send(Unknown Source:0)
        at kotlinx.coroutines.flow.internal.SendingCollector.emit(SendingCollector.kt:19)
        at kotlinx.coroutines.flow.FlowKt__ErrorsKt$catchImpl$$inlined$collect$1.emit(Collect.kt:137)
        at kotlinx.coroutines.flow.internal.SafeCollector.emit(SafeCollector.kt:33)
        at com.squareup.sample.dungeon.TickerWorker$run$1.invokeSuspend(GameWorkflow.kt:222)
        at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
        at kotlinx.coroutines.ResumeModeKt.resumeMode(ResumeMode.kt:67)
        at kotlinx.coroutines.DispatchedKt.resume(Dispatched.kt:319)
        at kotlinx.coroutines.DispatchedKt.dispatch(Dispatched.kt:301)
        at kotlinx.coroutines.CancellableContinuationImpl.dispatchResume(CancellableContinuationImpl.kt:250)
        at kotlinx.coroutines.CancellableContinuationImpl.resumeImpl(CancellableContinuationImpl.kt:260)
        at kotlinx.coroutines.CancellableContinuationImpl.resumeUndispatched(CancellableContinuationImpl.kt:332)
        at kotlinx.coroutines.android.HandlerContext$scheduleResumeAfterDelay$$inlined$Runnable$1.run(Runnable.kt:19)
        at android.os.Handler.handleCallback(Handler.java:873)
        at android.os.Handler.dispatchMessage(Handler.java:99)
        at android.os.Looper.loop(Looper.java:193)
        at android.app.ActivityThread.main(ActivityThread.java:6669)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:493)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:858)
```

**Comment by zach-klippenstein** [comment]
I can repro this even without minification enabled, so I'm guessing you're actually seeing a different exception that's being masked.

I'm playing with a new tracing API in the dungeon workflow, and (sometimes) seeing the same exception when I back out of the activity. It looks like an exception is actually being thrown from the builder itself (not one of its cases), and there's a bug in the coroutines library in that it tries reporting that exception to the continuation after it's already failed (maybe cancelled)?

**Comment by zach-klippenstein** [comment]
Was able to consistently reproduce this crash any time you get eaten, again without using R8.

Looks like what happens is the workers all get closed, and one of the `WorkflowNode`'s worker select clauses actually gets invoked twice. The second time, when the select statement tries resuming with the result of the clause, it throws because it was already resumed the first time.

I tried reproing in a simpler case, and I got a different exception but discovered that `onReceiveOrClosed` is _extremely_ broken. Filed a coroutines bug as https://github.com/Kotlin/kotlinx.coroutines/issues/1584, and I'm pretty sure that's what's causing this.
